geoserver: add service

rolfschr opened the pull request:
## Description of changes

This PR adds a new `services.geoserver` module which allows to run `geoserver` as a systemd service.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rolfschr committed:
geoserver: add service

rolfschr commented:
@imincik Could you have a look into this PR?

imincik commented on nixos/modules/services/web-apps/geoserver.nix:
```diff
@@ -0,0 +1,76 @@
+      };
+
+      serviceConfig = {
+        User = cfg.user;
```
* I suggest to use systemd's `DynamicUser` instead of real user/group.

See for example here: https://github.com/NixOS/nixpkgs/blob/d51f8579f03a76718d643dd184ae964e6f0dd751/nixos/modules/services/matrix/matrix-sliding-sync.nix#L96

* It is also worth to consider to use `StateDirectory` instead of manual `datadir` configuration.

rolfschr replied:
Ah, I considered both but refrained from it because I thought people might want to copy stuff into the data dir.
But fine with me, I'll adapt the service.

imincik replied:
@rolfschr , are you still interested in finishing this PR ?

imincik reviewed (changes requested):
Thanks for your work @rolfschr . I left some comments with suggestions.

rolfschr commented:
@imincik Sorry for the delay. Yes, I was still working on this. As per your idea to use `StateDirectory` I played around a lot to have a nice way to either allow the data dir to be stored in `/var/lib` (and thus use `StateDirectory`) *or* have it outside `/var/lib` (but then *not* use `StateDirectory`). Unfortunately, I haven't come up with a nice solution and now I have removed the possibility to choose the data dir location all together. (I had issues with permissions, and also had to look into workarounds like [this](https://github.com/NixOS/nixpkgs/blob/932390c76b7e0d44656a648fe336a706f659cd42/nixos/modules/services/continuous-integration/jenkins/default.nix#L239)).

I also wanted to have a mean to provide jetty args (say, change the port). Providing `-D...` as java opts doesn't work (IMHO). I have tried out several things, but here again, I am not satisfied and, at least for now, opted for not doing anything. For reference, this is how I changed the port on a (prod) server:

```nix
  geoserverWithExtensions = ...;
  geoserver = geoserverWithExtensions.overrideAttrs (finalAttrs: previousAttrs:
    let
      previousPostInstall = if builtins.hasAttr "postInstall" previousAttrs then previousAttrs.postInstall else "";
    in {
      postInstall = previousPostInstall + ''
        sed -i "s/jetty.http.port=8080/jetty.http.port=12345/" $out/share/geoserver/start.ini
        '';
    }
  );
```

This is not nice which is why I wanted to provide a better means when implementing `services.geoserver`. However, as I said, I haven't been able to find a good way.

Thus ... I just pushed my current version. It is now very bare bones and I'm not sure whether this is sufficient to justify a new `service`. I'm happy to further work on this but would need further guidance.

h7x4 commented:
@rolfschr I have seen some people use bind mounts to get around the datadir-not-in-var-lib issue in a nice way, but I'm not sure how well it plays together with `DynamicUser`. Maybe have a look at [`nixos/modules/services/audio/gonic.nix`](https://github.com/NixOS/nixpkgs/blob/e8d0a51d9aa7cf6bb72c3bcc2a106de2d2e7ad46/nixos/modules/services/audio/gonic.nix#L51-L69)? They seem to have gotten it working.

rolfschr commented:
@h7x4 Thanks for your input. I have been able to use a non-/var/lib directory as you suggested using bind mounts. This works as expected. The drawback I see here is that the log files will always print `/var/lib/geoserver` as data dir (because that's what the systemd unit thinks it is). Feel free to have a look into this.

As for the jetty configuration, I spent further time reading the docs and figured one can modify the jetty properties on the command line. However, a little patch is needed. I have added it in this PR. I'm not so happy that we now have a dedicated `*.patch` file as well as a `postPatch`-`sed`-based approach. I didn't want to add a patch file (yet).

I have now several commits here which I haven't squashed yet. Please have a look into this (also @imincik) and let me know what you all think.

rolfschr commented:
@imincik Could you have a look here, please? I'd like to finish this PR.

rolfschr commented:
I fixed the formatting but I think ofBorg didn't realize that. I assume there will be change requests so the checks will rerun anyways.

Please have a look into this now.

rolfschr committed:
tests/geoserver: disable check for GDAL

Loading the GDAL extension doesn't work as of 2024/10 due to problems
with XML. They seem to be the same as
https://github.com/NixOS/nixpkgs/pull/340707#issuecomment-2361894717
However, gdal / netcdf seem to be patched already so I don't understand
why this still fails. I hope this gets fixed in the future.

rolfschr commented:
@NixOS/geospatial Can you please look into this? I am waiting for feedback here. I had to resolve the merge conflict repeatedly because the rl-2411.section.md keeps getting updated. This PR stems from June.

Note that I have added an additional commit to disable the check for the gdal extension. It does in fact not work but I somehow believe/hope it will work auto-magically one day again. It seems to be related to libxml2 having disabled HTTP support by default. This seems to be [known](https://github.com/NixOS/nixpkgs/pull/340707#issuecomment-2361894717) issue in nixpkgs but I haven't managed to fix it. Actually, the problem [should not occur in the first place](https://github.com/NixOS/nixpkgs/blob/54cce652cb480258b12f15b2915e7a6c79ac51ae/pkgs/development/libraries/gdal/default.nix#L183) so I wonder what is going on. I tried to override in at other locations (gdal, netcdf) but to no avail. Maybe someone has a fix to propose but I do have to say I spent so much time already that I have now decided to simply disable the test for now (to make the build green) and keep an eye on this.

--- 

Here is an extract of the test log output:
```bash
vm-test-run-geoserver> Native library load failed.
vm-test-run-geoserver> java.lang.UnsatisfiedLinkError: /nix/store/qpi9ac8rkp51wbx87f38nj3chcd84h89-gdal-3.9.2/lib/jni/libgdalalljni.so: /nix/store/8627wpd4mbygb4p5ikzy7whvqhclqnc4-libspatialite-5.1.0/lib/libspatialite.so.8: undefined symbol: xmlNanoHTTPCleanup, version LIBXML2_2.4.30
vm-test-run-geoserver> Oct 12, 2024 1:54:14 PM it.geosolutions.imageio.gdalframework.GDALUtilities loadGDAL
vm-test-run-geoserver> WARNING: Failed to load the GDAL native libs. This is not a problem unless you need to use the GDAL plugins: they won't be enabled.
```

---

> As for the jetty configuration, I spent further time reading the docs and figured one can modify the jetty properties on the command line. However, a little patch is needed. I have added it in this PR. I'm not so happy that we now have a dedicated `*.patch` file as well as a `postPatch`-`sed`-based approach. I didn't want to add a patch file (yet).
> 
> I have now several commits here which I haven't squashed yet. Please have a look into this (also @imincik) and let me know what you all think.

Commits squashed and patch file added.

autra commented:
>  It seems to be related to libxml2 having disabled HTTP support by default.

This will be resolved soon I hope, with https://github.com/NixOS/nixpkgs/pull/350694. Maybe you can try to rebase this PR on that and see if it fixes it?

Apart from that the code looks good, but I didn't test and I'm not a geoserver user, so I'd prefer someone with more exp than me to approve.

> This PR stems from June.

Once, it took me a year and a half to get a PR merged (not on nixpkgs though), you still have a long way to go before beating this 😜
